server/src/test/java/org/elasticsearch/search/vectors/AbstractKnnVectorQueryBuilderTestCase.java

```diff
@@ -0,0 +1,245 @@
+        };
+        assertEquals(query, knnVectorQueryBuilt);
+    }
+
+    public void testWrongDimension() {
+        SearchExecutionContext context = createSearchExecutionContext();
+        KnnVectorQueryBuilder query = new KnnVectorQueryBuilder(VECTOR_FIELD, new float[] { 1.0f, 2.0f }, 10);
+        IllegalArgumentException e = expectThrows(IllegalArgumentException.class, () -> query.doToQuery(context));
+        assertThat(e.getMessage(), containsString("the query vector has a different dimension [2] than the index vectors [3]"));
+    }
+
+    public void testNonexistentField() {
+        SearchExecutionContext context = createSearchExecutionContext();
+        KnnVectorQueryBuilder query = new KnnVectorQueryBuilder("nonexistent", new float[] { 1.0f, 1.0f, 1.0f }, 10);
+        IllegalArgumentException e = expectThrows(IllegalArgumentException.class, () -> query.doToQuery(context));
+        assertThat(e.getMessage(), containsString("field [nonexistent] does not exist in the mapping"));
+    }
+
+    public void testWrongFieldType() {
+        SearchExecutionContext context = createSearchExecutionContext();
+        KnnVectorQueryBuilder query = new KnnVectorQueryBuilder(
+            AbstractBuilderTestCase.KEYWORD_FIELD_NAME,
+            new float[] { 1.0f, 1.0f, 1.0f },
+            10
+        );
+        IllegalArgumentException e = expectThrows(IllegalArgumentException.class, () -> query.doToQuery(context));
+        assertThat(e.getMessage(), containsString("[knn] queries are only supported on [dense_vector] fields"));
+    }
+
+    @Override
+    public void testValidOutput() {
+        KnnVectorQueryBuilder query = new KnnVectorQueryBuilder(VECTOR_FIELD, new float[] { 1.0f, 2.0f, 3.0f }, 10);
+        String expected = """
+            {
+              "knn" : {
+                "field" : "vector",
+                "vector" : [
+                  1.0,
+                  2.0,
+                  3.0
+                ],
+                "num_candidates" : 10
+              }
+            }""";
+        assertEquals(expected, query.toString());
+    }
+
+    @Override
+    public void testMustRewrite() throws IOException {
+        SearchExecutionContext context = createSearchExecutionContext();
+        context.setAllowUnmappedFields(true);
+        TermQueryBuilder termQuery = new TermQueryBuilder("unmapped_field", 42);
+        KnnVectorQueryBuilder query = new KnnVectorQueryBuilder(VECTOR_FIELD, new float[] { 1.0f, 2.0f, 3.0f }, VECTOR_DIMENSION);
+        query.addFilterQuery(termQuery);
+
+        IllegalStateException e = expectThrows(IllegalStateException.class, () -> query.toQuery(context));
+        assertEquals("Rewrite first", e.getMessage());
+
+        QueryBuilder rewrittenQuery = query.rewrite(context);
+        assertThat(rewrittenQuery, instanceOf(MatchNoneQueryBuilder.class));
+    }
+
+    public void testOldVersionSerialization() throws IOException {
```

Comment:
shall we specify the version in the name? Pretty sure old will not mean much in a couple of months.

Reply:
It is multi-version BWC, so, I will rename to not be "old"